stl/inc/algorithm

```diff
@@ -3988,6 +4267,92 @@ _NODISCARD bool is_heap(_ExPo&& _Exec, _RanIt _First, _RanIt _Last) noexcept /*
+    _NODISCARD constexpr _It _Is_heap_until_unchecked(
+        _It _First, const iter_difference_t<_It> _Size, _Pr _Pred, _Pj _Proj) {
+        // find extent of counted range that is a heap ordered by _Pred
+        _STL_INTERNAL_STATIC_ASSERT(random_access_iterator<_It>);
+        _STL_INTERNAL_STATIC_ASSERT(indirect_strict_weak_order<_Pr, projected<_It, _Pj>>);
+
+        if (_Size == 0) {
+            return _First;
+        }
+
+        using _Diff = iter_difference_t<_It>;
+        _Diff _Off  = 1;
+        for (; _Off < _Size; ++_Off) {
+            const auto _Mid1 = _First + static_cast<_Diff>((_Off - 1) >> 1); // shift for codegen
+            auto _Mid2       = _First + _Off;
+            if (_STD invoke(_Pred, _STD invoke(_Proj, *_Mid1), _STD invoke(_Proj, *_Mid2))) {
+                return _Mid2;
+            }
+        }
+
+        return _First + _Off;
+    }
+
+    class _Is_heap_fn : private _Not_quite_object {
+    public:
+        using _Not_quite_object::_Not_quite_object;
+
+        template <random_access_iterator _It, sentinel_for<_It> _Se, class _Pj = identity,
+            indirect_strict_weak_order<projected<_It, _Pj>> _Pr = ranges::less>
+        _NODISCARD constexpr bool operator()(_It _First, _Se _Last, _Pr _Pred = {}, _Pj _Proj = {}) const {
+            _Adl_verify_range(_First, _Last);
+            auto _UFirst      = _Get_unwrapped(_STD move(_First));
+            const auto _ULast = _Get_unwrapped(_STD move(_Last));
+            const auto _Size  = _RANGES distance(_UFirst, _ULast);
+            const auto _UResult =
+                _RANGES _Is_heap_until_unchecked(_STD move(_UFirst), _Size, _Pass_fn(_Pred), _Pass_fn(_Proj));
+            return _UResult == _ULast;
+        }
+
+        template <random_access_range _Rng, class _Pj = identity,
+            indirect_strict_weak_order<projected<iterator_t<_Rng>, _Pj>> _Pr = ranges::less>
+        _NODISCARD constexpr bool operator()(_Rng&& _Range, _Pr _Pred = {}, _Pj _Proj = {}) const {
+            const auto _Size = _RANGES distance(_Range);
+            const auto _UResult =
+                _RANGES _Is_heap_until_unchecked(_Ubegin(_Range), _Size, _Pass_fn(_Pred), _Pass_fn(_Proj));
+            return _UResult == _Uend(_Range);
+        }
+    };
+
+    inline constexpr _Is_heap_fn is_heap{_Not_quite_object::_Construct_tag{}};
+
+    // VARIABLE ranges::is_heap_until
+    class _Is_heap_until_fn : private _Not_quite_object {
+    public:
+        using _Not_quite_object::_Not_quite_object;
+
+        template <random_access_iterator _It, sentinel_for<_It> _Se, class _Pj = identity,
+            indirect_strict_weak_order<projected<_It, _Pj>> _Pr = ranges::less>
+        _NODISCARD constexpr _It operator()(_It _First, _Se _Last, _Pr _Pred = {}, _Pj _Proj = {}) const {
+            _Adl_verify_range(_First, _Last);
+            auto _UFirst     = _Get_unwrapped(_STD move(_First));
+            auto _ULast      = _Get_unwrapped(_STD move(_Last));

```

Comment:
> I assume it is because `_Is_heap_until_unchecked` is size based.

This precisely. We need the size, and could care less about the final iterator value, so `_Get_final_iterator_unwrapped` wouldn't make sense here. (Keeping `_ULast` around doesn't really make sense either - I should just pass `_Get_unwrapped(_STD move(_Last))` into the `distance` call.